### PR TITLE
🚨 Pass tp_plan from lm_heads

### PR DESCRIPTION
**Pull request by molbap**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47253)

# What does this PR do?

Noticed currently the `lm_head`-specific tp plans are wiped in `PreTrainedModel` post-init (we just ignore them). This should fix it

### CODE REVIEW

File: tests/test_tensor_parallel_mixin.py [hunk 1]
```diff
@@ -491,6 +491,15 @@ def _get_tp_model_class(self):
...
         return self.all_model_classes[0]
 
+    def _get_tp_config(self):
+        """Tiny config with `vocab_size` rounded up to a multiple of the world size, as sharded dims (typically `lm_head`) have to be split across ranks."""
+        config = self.model_tester.get_config()
+        text_config = config.get_text_config()
+        remainder = text_config.vocab_size % self.tensor_parallel_size
+        if remainder:
+            text_config.vocab_size += self.tensor_parallel_size - remainder
+        return config
```
Comment (molbap) [inline tests/test_tensor_parallel_mixin.py:501]: this is because we have typically vocab size 99 on tiny tests, which can't be sharded on 2 ranks obviously

Comment (vasqu) [reply tests/test_tensor_parallel_mixin.py:501]: I'm thinking wheter we should automatically extend an embedding if we notice this? So in from pretrained, if we notice 
1. TP
2. Embeddings that get shared we resize the embeddings before

Comment (molbap) [reply tests/test_tensor_parallel_mixin.py:501]: ah auto-pad you mean? hmm we could ig, I feel like we are already doing it in a couple other places IIRC

Comment (vasqu) [reply tests/test_tensor_parallel_mixin.py:501]: Yep, because users will complain I'm sure :sweat_smile: Maybe a warning but having users to think leads to problems

Comment (molbap) [reply tests/test_tensor_parallel_mixin.py:501]: but we should pad just before the gather and unpad right after, no?

Comment (molbap) [reply tests/test_tensor_parallel_mixin.py:501]: other idea (sorry for the noise): we should just validate, feels less magic. like if TP is provided with a wrong vocab size just raise from the get-go. Should be in another PR though because it could break existing setups

Comment (vasqu) [reply tests/test_tensor_parallel_mixin.py:501]: I think this is already breaking in itself no? Users could have arbitrary vocab sizes before this

But yea, I can see the validation path - less magic and add how to properly do it

Comment (molbap) [reply tests/test_tensor_parallel_mixin.py:501]: ah yes, you're right... before it was silently replicating so of course this will raise. I'll add the validation logic then

**Review by vasqu — APPROVED** [review]
Imo, this looks good but let's wait for @ArthurZucker and @3outeille for a final look 🫡

**Review by 3outeille — APPROVED** [review]

File: src/transformers/integrations/tensor_parallel.py [hunk 2]
```diff
@@ -719,6 +723,16 @@ def __init__(self, gather_output: bool = False, **kwargs):
...
+    def validate_module(self, module: nn.Module, device_mesh, layer_name: str = ""):
+        out_features = getattr(module, "out_features", None)
+        if self.gather_output and out_features is not None and out_features % device_mesh.size() != 0:
+            raise ValueError(
+                f"`{layer_name}` ({type(module).__name__} with out_features={out_features}) is sharded with "
+                f"'colwise_gather_output', which requires out_features to be divisible by the number of ranks "
+                f"({device_mesh.size()}) to all-gather equal-size shards. Resize the weight (e.g. "
+                f"`model.resize_token_embeddings` for LM heads) or override this module's entry in the tp_plan."
+            )
+
```
Comment (ArthurZucker) [inline src/transformers/integrations/tensor_parallel.py:735]: since only this layer uses it, we can not make it generic for now

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -1415,15 +1418,18 @@ def post_init(self):
         """
         # Attach the different parallel plans and tied weight keys to the top-most model, so that everything is
         # easily available.
-        self._tp_plan, self._ep_plan, self._pp_plan, self._fsdp_plan = {}, {}, {}, {}
+        # Start from the class-level plans (e.g. `{"lm_head": "colwise_rep"}` on `...ForCausalLM` classes), copying
+        # them as they are mutated below and would otherwise contaminate the class attribute shared by all instances
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1422]: absolutely, this was already the case.... do you know where the regression comes from?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1268,6 +1268,9 @@ class PreTrainedModel(nn.Module, EmbeddingAccessMixin, ModuleUtilsMixin, PushToH
     # models, this attribute is currently defined in respective model code. For base models, it comes from
     # `config.base_model_pp_plan` during `post_init`.
     _pp_plan: dict[str, tuple[str, str]] = None
+    # An expert parallel plan used instead of `_tp_plan` when expert parallelism is enabled. For base models, it comes
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1271]: not super

File: tests/test_tensor_parallel_mixin.py (hunk 1 above)
Comment (ArthurZucker) [reply tests/test_tensor_parallel_mixin.py:501]: IMO, if it does not cost much and we can resize for the user, would be nice. People probably expect us to do so, and we are the ones setting the plan to default to colwise rep. 
The main concern is for the gather to not take into account the padding, which is important

**Review by ArthurZucker — COMMENTED** [review]
ty, not sure we need generic, and as pointed let's maybe pad by default?
we can isolate the 2 changes. let's also trace the blame on PR that regressed the init, it was I am certain working before

File: src/transformers/integrations/tensor_parallel.py (hunk 2 above)
Comment (molbap) [reply src/transformers/integrations/tensor_parallel.py:735]: yeah it's a bit awkward. Wdy suggest? move it the the layer?

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (molbap) [reply src/transformers/modeling_utils.py:1422]: yeah wrote that mostly for review. Yes, it seems to be coming from a big Revert, likely an oversight https://github.com/huggingface/transformers/issues/46246

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (molbap) [reply src/transformers/modeling_utils.py:1271]: can you elaborate? :grin:

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (3outeille) [reply src/transformers/modeling_utils.py:1422]: it's actually in #36677 and then we keep the pattern

Comment (molbap) [reply src/transformers/modeling_utils.py:1422]: Are you sure @3outeille ? it doesn't seem to be there that we set 
` self._tp_plan, self._ep_plan, self._pp_plan, self._fsdp_plan = {}, {}, {}, {}`

**Comment by ArthurZucker** [comment]
Merging we need it for release
